multiversx_sdk/abi/serializer.py

```diff
@@ -0,0 +1,116 @@
+from multiversx_sdk.abi.parts import PartsHolder
+from multiversx_sdk.abi.values_multi import *
+
+
+class Serializer:
+    """
+    The serializer follows the rules of the MultiversX Serialization format:
+    https://docs.multiversx.com/developers/data/serialization-overview
+    """
+
+    def __init__(self, parts_separator: str):
+        if not parts_separator:
+            raise ValueError("cannot create serializer: parts separator must not be empty")
+
+        self.parts_separator = parts_separator
+        self.codec = Codec()
+
+    def serialize(self, input_values: Sequence[Any]) -> str:
+        parts = self.serialize_to_parts(input_values)
+        return self._encode_parts(parts)
+
+    def serialize_to_parts(self, input_values: Sequence[Any]) -> List[bytes]:
+        parts_holder = PartsHolder([])
+        self._do_serialize(parts_holder, input_values)
+        return parts_holder.get_parts()
+
+    def _do_serialize(self, parts_holder: PartsHolder, input_values: Sequence[Any]):
+        for i, value in enumerate(input_values):
+            if value is None:
+                raise ValueError("cannot serialize nil value")
+
+            if isinstance(value, OptionalValue):
+                if i != len(input_values) - 1:
+                    # Usage of multiple optional values is not recommended:
+                    # https://docs.multiversx.com/developers/data/multi-values
+                    # Thus, here, we disallow them.
+                    raise ValueError("an optional value must be last among input values")
+
+                if value.value is not None:
+                    self._do_serialize(parts_holder, [value.value])
+            elif isinstance(value, MultiValue):
+                self._do_serialize(parts_holder, value.items)
+            elif isinstance(value, VariadicValues):
+                if i != len(input_values) - 1:
+                    raise ValueError("variadic values must be last among input values")
+
+                self._do_serialize(parts_holder, value.items)
+            else:
+                parts_holder.append_empty_part()
+                self._serialize_single_value(parts_holder, value)
+
+    def _serialize_single_value(self, parts_holder: PartsHolder, value: Any):
+        data = self.codec.encode_top_level(value)
+        parts_holder.append_to_last_part(data)
+
+    def deserialize(self, data: str, output_values: Sequence[Any]):
+        parts = self._decode_into_parts(data)
+        self.deserialize_parts(parts, output_values)
+
+    def deserialize_parts(self, parts: Sequence[bytes], output_values: Sequence[Any]):
+        parts_holder = PartsHolder(parts)
+        self._do_deserialize(parts_holder, output_values)
+
+    def _do_deserialize(self, parts_holder: PartsHolder, output_values: Sequence[Any]):
+        for i, value in enumerate(output_values):
+            if value is None:
+                raise ValueError("cannot deserialize into nil value")
```

Comment:
same comm as above regarding `nil`

Reply:
Fixed in https://github.com/multiversx/mx-sdk-py/pull/49.